In [1]:
import pandas as pd
import numpy as np

In [2]:
import os, sys

In [3]:
import tensorflow
import tensorflow as tf

In [8]:
# sanity check for importing tensorflow (from virtualenv)
print(sys.executable)
sys.path

/home/mobicfd/ReacCFD/tf/bin/python2


['',
 '/home/mobicfd/ReacCFD/tf/lib/python2.7',
 '/home/mobicfd/ReacCFD/tf/lib/python2.7/plat-x86_64-linux-gnu',
 '/home/mobicfd/ReacCFD/tf/lib/python2.7/lib-tk',
 '/home/mobicfd/ReacCFD/tf/lib/python2.7/lib-old',
 '/home/mobicfd/ReacCFD/tf/lib/python2.7/lib-dynload',
 '/usr/lib/python2.7',
 '/usr/lib/python2.7/plat-x86_64-linux-gnu',
 '/usr/lib/python2.7/lib-tk',
 '/home/mobicfd/ReacCFD/tf/local/lib/python2.7/site-packages',
 '/home/mobicfd/.local/lib/python2.7/site-packages',
 '/usr/local/lib/python2.7/site-packages',
 '/usr/local/lib/python2.7/dist-packages',
 '/usr/lib/python2.7/dist-packages',
 '/usr/lib/python2.7/dist-packages/PILcompat',
 '/usr/lib/python2.7/dist-packages/gtk-2.0',
 '/usr/lib/python2.7/dist-packages/wx-3.0-gtk2',
 '/home/mobicfd/ReacCFD/tf/local/lib/python2.7/site-packages/IPython/extensions',
 '/home/mobicfd/.ipython']

In [9]:
os.listdir( os.getcwd()  )

['departments.csv.zip',
 'products.csv.zip',
 'sample_submission.csv',
 'products.csv',
 'order_products__train.csv.zip',
 'orders.csv.zip',
 '.ipynb_checkpoints',
 'aisles.csv',
 'aisles.csv.zip',
 'orders.csv',
 'order_products__train.csv',
 'InstacartMarketBasket.ipynb',
 'order_products__prior.csv.zip',
 'sample_submission.csv.zip',
 'departments.csv',
 'order_products__prior.csv']

In [11]:
aislesDF = pd.read_csv(os.getcwd() +"/aisles.csv")
departmentsDF = pd.read_csv(os.getcwd() +"/departments.csv")
productsDF = pd.read_csv(os.getcwd() +"/products.csv")
order_products__trainDF = pd.read_csv(os.getcwd() +"/order_products__train.csv")
ordersDF = pd.read_csv(os.getcwd() +"/orders.csv")
order_products__priorDF = pd.read_csv(os.getcwd() +"/order_products__prior.csv")

In [12]:
print(aislesDF.columns)
print(departmentsDF.columns)
print(ordersDF.columns)
print(order_products__priorDF.columns)
print(order_products__trainDF.columns)
print(productsDF.columns)

print(aislesDF.describe())
print(departmentsDF.describe())
print(ordersDF.describe())
print(order_products__priorDF.describe())
print(order_products__trainDF.describe())
print(productsDF.describe())


Index([u'aisle_id', u'aisle'], dtype='object')
Index([u'department_id', u'department'], dtype='object')
Index([u'order_id', u'user_id', u'eval_set', u'order_number', u'order_dow',
       u'order_hour_of_day', u'days_since_prior_order'],
      dtype='object')
Index([u'order_id', u'product_id', u'add_to_cart_order', u'reordered'], dtype='object')
Index([u'order_id', u'product_id', u'add_to_cart_order', u'reordered'], dtype='object')
Index([u'product_id', u'product_name', u'aisle_id', u'department_id'], dtype='object')
         aisle_id
count  134.000000
mean    67.500000
std     38.826537
min      1.000000
25%     34.250000
50%     67.500000
75%    100.750000
max    134.000000
       department_id
count      21.000000
mean       11.000000
std         6.204837
min         1.000000
25%         6.000000
50%        11.000000
75%        16.000000
max        21.000000
           order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06 

## `aisles`

In [13]:
print(aislesDF.head())
print(aislesDF['aisle_id'].unique())
print(len( aislesDF['aisle_id'].unique()) )
print( aislesDF['aisle'].unique()[:10] ) 
print(len(aislesDF['aisle'].unique()) )


   aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation
[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134]
134
['prepared soups salads' 'specialty cheeses' 'energy granola bars'
 'instant foods' 'marinades meat preparation' 'other' 'packaged meat'
 'bakery desserts' 'pasta sauce' 'kitchen supplies']
134


And so clear `aisle` describes the 134 unique aisles that each product belongs too.  

## departments

In [14]:
print(departmentsDF.head())
print(departmentsDF['department_id'].unique()[:10])
print(len( departmentsDF['department_id'].unique()) )
print( departmentsDF['department'].unique()[:10] ) 
print(len(departmentsDF['department'].unique()) )

   department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol
[ 1  2  3  4  5  6  7  8  9 10]
21
['frozen' 'other' 'bakery' 'produce' 'alcohol' 'international' 'beverages'
 'pets' 'dry goods pasta' 'bulk']
21


## `products`  

In [15]:
print(productsDF.head())
print(productsDF['product_id'].unique()[:10])
print(len( productsDF['product_id'].unique()) )
print( productsDF['product_name'].unique()[:10] ) 
print(len(productsDF['product_name'].unique()) )
print(productsDF['aisle_id'].unique()[:10])
print(len( productsDF['aisle_id'].unique()) )
print(productsDF['department_id'].unique()[:10])
print(len( productsDF['department_id'].unique()) )


   product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13  
[ 1  2  3  4  5  6  7  8  9 10]
49688
['Chocolate Sandwich Cookies' 'All-Seasons Salt'
 'Robust Golden Unsweetened Oolong Tea'
 'Smart Ones Classic Favorites Mini Rigatoni With Vodka Cream Sauce'
 'Green Chile Anytime Sauce' 'Dry Nose Oil'
 'Pure Coconut Water With Orange' "Cut Russet Potatoes Steam N' Mash"
 'Light Strawberry Blueberry Yogurt'
 'Sparkling Orange Juice & Prickly Pear Beverage']
49688
[ 61 104  94 

## `order_product_*`  

cf. [`instacart-market-basket-analysis/data`](https://www.kaggle.com/c/instacart-market-basket-analysis/data)

"These files specify which products were purchased in each order.  
* `order_products__prior.csv` contains previous order contents for all customers. 
* 'reordered' indicates that the customer has a previous order that contains the product.

Note that some orders will have no reordered items. You may predict an explicit 'None' value for orders with no reordered items. "


In [16]:
order_products__priorDF.columns

Index([u'order_id', u'product_id', u'add_to_cart_order', u'reordered'], dtype='object')

In [17]:
print(order_products__priorDF.head())
print(order_products__priorDF['order_id'].unique()[:10])
print(len( order_products__priorDF['order_id'].unique()) )
print( order_products__priorDF['product_id'].unique()[:10] ) 
print(len( order_products__priorDF['product_id'].unique()) )
print(order_products__priorDF['add_to_cart_order'].unique()[:10])
print(len( order_products__priorDF['add_to_cart_order'].unique()) )
print(order_products__priorDF['reordered'].unique()[:10])
print(len( order_products__priorDF['reordered'].unique()) )

   order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0
[ 2  3  4  5  6  7  8  9 10 11]
3214874
[33120 28985  9327 45918 30035 17794 40141  1819 43668 33754]
49677
[ 1  2  3  4  5  6  7  8  9 10]
145
[1 0]
2


In [18]:
print( order_products__priorDF.loc[order_products__priorDF['order_id']==2].columns )
order_products__priorDF.loc[order_products__priorDF['order_id']==2]

Index([u'order_id', u'product_id', u'add_to_cart_order', u'reordered'], dtype='object')


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0
5         2       17794                  6          1
6         2       40141                  7          1
7         2        1819                  8          1
8         2       43668                  9          0

In [19]:
order_id_2 = 2
print(order_products__priorDF.loc[order_products__priorDF['order_id']==2].columns)
prodseq_2    = (order_products__priorDF.loc[order_products__priorDF['order_id']==2])['product_id']
reorderseq_2 = (order_products__priorDF.loc[order_products__priorDF['order_id']==2])['reordered']
print(prodseq_2.dtype)

Index([u'order_id', u'product_id', u'add_to_cart_order', u'reordered'], dtype='object')
int64


In [20]:
print(prodseq_2.values)
print(reorderseq_2.values)
zip(prodseq_2.values,reorderseq_2.values)

[33120 28985  9327 45918 30035 17794 40141  1819 43668]
[1 1 0 1 0 1 1 1 0]


[(33120, 1),
 (28985, 1),
 (9327, 0),
 (45918, 1),
 (30035, 0),
 (17794, 1),
 (40141, 1),
 (1819, 1),
 (43668, 0)]

In [21]:
print(order_products__trainDF.head())
print(order_products__trainDF['order_id'].unique()[:10])
print(len( order_products__trainDF['order_id'].unique()) )
print( order_products__trainDF['product_id'].unique()[:10] ) 
print(len( order_products__trainDF['product_id'].unique()) )
print(order_products__trainDF['add_to_cart_order'].unique()[:10])
print(len( order_products__trainDF['add_to_cart_order'].unique()) )
print(order_products__trainDF['reordered'].unique()[:10])
print(len( order_products__trainDF['reordered'].unique()) )

   order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1
[  1  36  38  96  98 112 170 218 226 349]
131209
[49302 11109 10246 49683 43633 13176 47209 22035 39612 19660]
39123
[ 1  2  3  4  5  6  7  8  9 10]
80
[1 0]
2


# `orders`  

This file tells to which set (prior, train, test) an order belongs. You are predicting reordered items only for the test set orders. 'order_dow' is the day of week.  

Order number is for each user (labeled with user_id).  Say this is a particular user's xth order.  Then order_number=x.  

In [28]:
print(ordersDF.head())
print(ordersDF.columns)
print(ordersDF.describe())
print(len(ordersDF['order_id'].unique()))
print(len(ordersDF['user_id'].unique()))
print((ordersDF['eval_set'].unique()))
print(len(ordersDF['order_number'].unique()))


   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
Index([u'order_id', u'user_id', u'eval_set', u'order_number', u'order_dow',
       u'order_hour_of_day', u'days_since_prior_order'],
      dtype='object')
           order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05

In [34]:
orders_prior = ordersDF.loc[ordersDF['eval_set']=='prior']
orders_train = ordersDF.loc[ordersDF['eval_set']=='train']
orders_test  = ordersDF.loc[ordersDF['eval_set']=='test']
orders_prior_train = ordersDF.loc[(ordersDF['eval_set']=='prior') | (ordersDF['eval_set']=='train')]


In [49]:
orders_prior_train_user_id=orders_prior_train.groupby(orders_prior_train['user_id'])
print( list( orders_prior_train_user_id )[:2])

[(1,     order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  


In [52]:
orders_prior_train_user_id.reset_index()


AttributeError: Cannot access callable attribute 'reset_index' of 'DataFrameGroupBy' objects, try using the 'apply' method

In [ ]:
orders_prior_train_uid.head()

In [48]:
#orders_prior_train_user_id_vec=list(zip(list(zip(orders_prior_train_user_id['order_dow'],orders_prior_train_user_id['order_hour_of_day'])), 
#        orders_prior_train_user_id['days_since_prior_order']))
#print(type(orders_prior_train_user_id_vec))
orders_prior_train_user_id['dow_hour'] = list(zip(orders_prior_train_uid['order_dow'],orders_prior_train_user_id['order_hour_of_day']))


TypeError: 'DataFrameGroupBy' object does not support item assignment

In [47]:
orders_prior_train_user_id.head()

0  \
0             (1, [2, 3, 3, 4, 4, 2, 1, 1, 1, 4, 4])   
1  (2, [2, 5, 1, 2, 3, 2, 2, 1, 2, 1, 1, 1, 4, 3,...   
2          (3, [1, 3, 3, 2, 0, 0, 0, 0, 0, 3, 0, 1])   

                                                   1  
0         (1, [8, 7, 12, 7, 15, 7, 9, 14, 16, 8, 8])  
1  (2, [11, 10, 10, 10, 11, 9, 12, 15, 9, 11, 10,...  
2  (3, [14, 19, 16, 18, 17, 16, 15, 17, 16, 16, 1...